aldroid

```diff
@@ -0,0 +1,754 @@
+    # We should stop this here instead of catching it later in the loop
+
+    # Determine output directory.
+    local -r OUTPUT_DIRECTORY=${2-/tmp/$1}
+
+    # Create output directory if it does not exist.
+    if [ ! -d "$OUTPUT_DIRECTORY" ]; then
+        mkdir "$OUTPUT_DIRECTORY"
+    fi
+
+    # Iterate over all OBBs we found and pull them from the device.
+    for OBB in "$TEMP_OBBS"; do
+        # In case there are other files there make sure the obb files end in .obb
+        if [[ "${OBB: -4}" != ".obb" ]]; then
+            continue
+        fi
+
+        adb pull "/sdcard/Android/obb/$1/$OBB" "$OUTPUT_DIRECTORY/$OBB"
+        echo "Pulled $OBB into $OUTPUT_DIRECTORY/$OBB"
+        OBBS+=$OBB
+    done
+
+    return 0
+}
+
+# Pushes OBBs back to the device. Uses array stored in OBBS variable and assumes they reside in /tmp/<package name>.
+# First argument is the package name for which to push OBBs for.
+push_obbs() {
+    # Verify an argument was provided
+    if [ -z "$1" ]; then
+        echo "No package name provided to push_obbs"
+        exit 1
+    fi
+
+    echo "Pushing OBB(s)..."
+    echo "$OBBS"
+
+    # Safety for empty arrays as if it's empty we would push the entire folder
+    if [[ ${#OBBS[@]} -eq 0 ]]; then
+        echo "No OBBs to push"
+        return 0
+    fi
+
+    # Iterate all OBBs on $OBBS and push them back to the device.
+    for OBB in "$OBBS"; do
+        adb push "/tmp/$1/$OBB" "/sdcard/Android/obb/$1/$OBB"
+    done
+
+    return 0
+}
+
+# Starts an app on the device given a package name.
+# First argument is the package name to start.
+start_app() {
+    # Verify an argument was provided
+    if [ -z $1 ]; then
+        echo "No package name provided to run_app"
+        exit 1
+    fi
+
+    local PACKAGE_NAME=$1
+
+    # monkey is a tool to simulate user interaction. In this case it simulates a press on the app icon.
+    adb shell monkey -p "${PACKAGE_NAME}" -c android.intent.category.LAUNCHER 1 >/dev/null 2>&1
+}
+
+# Parses an AndroidManifest.xml to extract the package name.
+# First argument is the path to the AndroidManifest.xml
+# Saves package name in EXTRACTED_PACKAGE_NAME
+extract_package_name_from_manifest() {
+    # Verify an argument was provided
+    if [ -z $1 ]; then
+        echo "No argument provided to extract_package_name_from_manifest"
+        exit 1
+    fi
+
+    if [ ! -f $1 ]; then
+        echo "$1 is not a path to a valid file"
+        exit 1
+    fi
+
+    # Get AndroidManifest content
+    local MANIFEST=$(cat "$1")
+    local PACKAGE_NAME=
+
+    # The manifest will include package inside the manifest tag.
+    # ex: <manifest package="com.my.package.name">
+
+    # Trim start until package="
+    PACKAGE_NAME=${MANIFEST#*package=\"}
+    # Trim everything after first "
+    PACKAGE_NAME=${PACKAGE_NAME%%\"*}
+
+    # Check if we have a valid package name
+    # TODOTODOTODO Guess this makes no sense and we should rather check if it contains any spaces or something
```

Comment:
Does this comment need an update? or is it still a TODO

Reply:
The comment needed an update, still a todo but I hope the newer TODO is clearer. Thinking about adding a regex check to validate that the final string in PACKAGE_NAME is something that looks like a package name (letters, numbers and periods only).